File: backend/phase2/zone_mapper.py
def canvas_objects_to_overlay_layers(
    canvas_objects: list[dict],
    field_configs: dict[str, dict],
    img_width: int,
    img_height: int,
    display_width: int,
    display_height: int
) -> list[dict]:
    """
    Converts streamlit-drawable-canvas rectangles to overlay_layers.
    Scales display coordinates to actual PNG pixels.
    """
    scale_x = img_width / display_width
    scale_y = img_height / display_height
    layers = []

    for i, obj in enumerate(canvas_objects):
        if obj.get("type") != "rect":
            continue

        cfg = field_configs.get(str(i), {})
        field_id = (
            cfg.get("id", f"field_{i + 1}")
            .strip()
            .replace(" ", "_")
            .lower()
        )
        field_type = cfg.get("type", "text")

        box = {
            "x": max(0, round(obj["left"] * scale_x)),
            "y": max(0, round(obj["top"] * scale_y)),
            "width": round(obj["width"] * scale_x),
            "height": round(obj["height"] * scale_y),
        }

        if field_type == "text":
            layer = {
                "id": field_id,
                "type": "text",
                "placeholder": f"{{{{{field_id}}}}}",
                "instruction": cfg.get(
                    "instruction", f"Value for {field_id}"
                ),
                "llm_can_invent": cfg.get("llm_can_invent", False),
                "box": box,
                "style": {
                    "font_family": cfg.get("font_family", "Poppins"),
                    "font_size": int(cfg.get("font_size", 48)),
                    "font_size_min": int(cfg.get("font_size_min", 20)),
                    "font_weight": cfg.get("font_weight", "bold"),
                    "font_style": cfg.get("font_style", "normal"),
                    "color": cfg.get("color", "#FFFFFF"),
                    "align": cfg.get("align", "center"),
                    "vertical_align": "middle",
                },
                "editable": True,
            }
        else:
            layer = {
                "id": field_id,
                "type": "image",
                "placeholder": f"{{{{{field_id}}}}}",
                "instruction": cfg.get(
                    "instruction", "Upload image to place here"
                ),
                "llm_can_invent": False,
                "box": box,
                "style": {
                    "shape": cfg.get("shape", "circle"),
                    "border_color": cfg.get("border_color", "#FFFFFF"),
                    "border_width": int(cfg.get("border_width", 4)),
                },
                "editable": True,
            }

        layers.append(layer)

    return layers


def validate_layers(
    layers: list[dict],
    canvas_w: int,
    canvas_h: int
) -> list[str]:
    """Returns list of error strings. Empty = all good."""
    errors = []
    ids_seen = set()

    for i, layer in enumerate(layers):
        fid = layer.get("id", "")

        if not fid:
            errors.append(f"Box {i + 1} has no field ID")
        elif fid in ids_seen:
            errors.append(f"Duplicate field ID: '{fid}'")
        else:
            ids_seen.add(fid)

        box = layer.get("box", {})
        if box.get("width", 0) < 20 or box.get("height", 0) < 10:
            errors.append(
                f"Field '{fid}' box is too small — "
                f"draw a larger rectangle"
            )

        if box.get("x", 0) + box.get("width", 0) > canvas_w:
            errors.append(
                f"Field '{fid}' extends past right edge of image"
            )
        if box.get("y", 0) + box.get("height", 0) > canvas_h:
            errors.append(
                f"Field '{fid}' extends past bottom edge of image"
            )

        if layer.get("type") == "text":
            s = layer.get("style", {})
            if s.get("font_size_min", 999) >= s.get("font_size", 0):
                errors.append(
                    f"Field '{fid}': min font size must be "
                    f"smaller than font size"
                )

    return errors

File: rag-template-generation-web/lib/zoneMapper.ts
export interface LayerBox {
  x: number;
  y: number;
  width: number;
  height: number;
}

export interface LayerStyle {
  font_family?: string;
  font_size?: number;
  font_size_min?: number;
  font_weight?: string;
  font_style?: string;
  color?: string;
  align?: string;
  vertical_align?: string;
  shape?: string;
  border_color?: string;
  border_width?: number;
}

export interface OverlayLayer {
  id: string;
  type: "text" | "image";
  placeholder: string;
  instruction: string;
  llm_can_invent: boolean;
  box: LayerBox;
  style: LayerStyle;
  editable: boolean;
}

export interface CanvasObject {
  type?: string;
  left: number;
  top: number;
  width: number;
  height: number;
  [key: string]: any;
}

export interface ZoneConfig {
  id?: string;
  type?: "text" | "image";
  instruction?: string;
  llm_can_invent?: boolean;
  font_family?: string;
  font_size?: number;
  font_size_min?: number;
  font_weight?: string;
  font_style?: string;
  color?: string;
  align?: string;
  shape?: string;
  border_color?: string;
  border_width?: number;
}

/**
 * Converts drawable canvas rectangles to overlay_layers.
 * Scales display coordinates (e.g. 480x640) to actual PNG pixels (e.g. 1024x1536).
 */
export function canvasObjectsToOverlayLayers(
  canvasObjects: CanvasObject[],
  fieldConfigs: Record<string, ZoneConfig>,
  imgWidth: number,
  imgHeight: number,
  displayWidth: number,
  displayHeight: number
): OverlayLayer[] {
  const scaleX = imgWidth / displayWidth;
  const scaleY = imgHeight / displayHeight;
  const layers: OverlayLayer[] = [];

  canvasObjects.forEach((obj, i) => {
    // Only process rectangular zones
    if (obj.type && obj.type !== "rect" && obj.type !== "Rect") {
      return;
    }

    const cfg = fieldConfigs[String(i)] || fieldConfigs[obj.id || ""] || {};
    const rawId = cfg.id || obj.id || `field_${i + 1}`;
    const fieldId = rawId.trim().replace(/\s+/g, "_").toLowerCase();
    const fieldType = (cfg.type || obj.type || "text") === "image" ? "image" : "text";

    const box: LayerBox = {
      x: Math.max(0, Math.round(obj.left * scaleX)),
      y: Math.max(0, Math.round(obj.top * scaleY)),
      width: Math.round(obj.width * scaleX),
      height: Math.round(obj.height * scaleY),
    };

    if (fieldType === "text") {
      layers.push({
        id: fieldId,
        type: "text",
        placeholder: `{{${fieldId}}}`,
        instruction: (cfg.instruction !== undefined && cfg.instruction.trim() !== "") ? cfg.instruction.trim() : `Value for ${fieldId}`,
        llm_can_invent: cfg.llm_can_invent ?? false,
        box,
        style: {
          font_family: cfg.font_family || "Poppins",
          font_size: Number(cfg.font_size ?? 48),
          font_size_min: Number(cfg.font_size_min ?? 20),
          font_weight: cfg.font_weight || "bold",
          font_style: cfg.font_style || "normal",
          color: cfg.color || "#FFFFFF",
          align: cfg.align || "center",
          vertical_align: "middle",
        },
        editable: true,
      });
    } else {
      layers.push({
        id: fieldId,
        type: "image",
        placeholder: `{{${fieldId}}}`,
        instruction: (cfg.instruction !== undefined && cfg.instruction.trim() !== "") ? cfg.instruction.trim() : "Upload image to place here",
        llm_can_invent: false,
        box,
        style: {
          shape: cfg.shape || "circle",
          border_color: cfg.border_color || "#FFFFFF",
          border_width: Number(cfg.border_width ?? 4),
        },
        editable: true,
      });
    }
  });

  return layers;
}

/**
 * Validates overlay layers according to size, ID, font, and canvas boundary rules.
 * Returns array of error string descriptions. Empty array = valid.
 */
export function validateLayers(
  layers: OverlayLayer[],
  canvasW: number,
  canvasH: number
): string[] {
  const errors: string[] = [];
  const idsSeen = new Set<string>();

  layers.forEach((layer, i) => {
    const fid = layer.id || "";

    if (!fid) {
      errors.push(`Zone #${i + 1} has no field ID.`);
    } else if (idsSeen.has(fid)) {
      errors.push(`Duplicate field ID: '${fid}'. Each zone must have a unique ID.`);
    } else {
      idsSeen.add(fid);
    }

    const box = layer.box || { x: 0, y: 0, width: 0, height: 0 };
    if (box.width < 20 || box.height < 10) {
      errors.push(`Field '${fid}' box is too small — draw a larger rectangle.`);
    }

    if (box.x + box.width > canvasW) {
      errors.push(`Field '${fid}' extends past right edge of image (${box.x + box.width}px > ${canvasW}px).`);
    }
    if (box.y + box.height > canvasH) {
      errors.push(`Field '${fid}' extends past bottom edge of image (${box.y + box.height}px > ${canvasH}px).`);
    }

    if (layer.type === "text") {
      const s = layer.style || {};
      const minFont = Number(s.font_size_min ?? 20);
      const mainFont = Number(s.font_size ?? 48);
      if (minFont >= mainFont) {
        errors.push(`Field '${fid}': minimum font size (${minFont}) must be smaller than main font size (${mainFont}).`);
      }
    }
  });

  return errors;
}
